src/julia.h

```diff
@@ -338,29 +376,41 @@ typedef struct _jl_module_t {
     uint32_t counter;
 } jl_module_t;
 
-typedef struct _jl_methlist_t {
+// one TupleType-to-Value entry
+typedef struct _jl_typemap_entry_t {
     JL_DATA_TYPE
-    jl_tupletype_t *sig;
-    int8_t va;
-    int8_t isstaged;
-    jl_svec_t *tvars;
-    jl_lambda_info_t *func;
-    // cache of specializations of this method for invoke(), i.e.
-    // cases where this method was called even though it was not necessarily
-    // the most specific for the argument types.
-    struct _jl_methtable_t *invokes;
-    // TODO: pointer from specialized to original method
-    //jl_function_t *orig_method;
-    struct _jl_methlist_t *next;
-} jl_methlist_t;
+    struct _jl_typemap_entry_t *next; // invasive linked list
+    jl_tupletype_t *sig; // the type signature for this entry
+    jl_svec_t *tvars; // the bound type variables for sig
+    jl_tupletype_t *simplesig; // a simple signature for fast rejection
+    jl_svec_t *guardsigs;
+    union {
+        jl_value_t *value;
+        jl_lambda_info_t *linfo;
+        jl_method_t *method;
+    } func; // [nullable]
+    // memoized properties of sig:
+    int8_t isleafsig; // isleaftype(sig) & !any(isType, sig) : unsorted and very fast
+    int8_t issimplesig; // all(isleaftype | isAny | isType | isVararg, sig) : sorted and fast
+    int8_t va; // isVararg(sig)
+} jl_typemap_entry_t;
+
+// one level in a TupleMap tree
+// indexed by key if it is a sublevel in an array
+typedef struct _jl_typemap_level_t {
+    JL_DATA_TYPE
+    jl_array_t *arg1; // Array{union jl_typemap_t}
+    jl_array_t *targ; // Array{union jl_typemap_t}
+    jl_typemap_entry_t *linear; // union jl_typemap_t (but no more levels)
+    jl_value_t *key; // [nullable]
```

Comment:
yes (taking inspiration from Glib, which uses this notation in its documentation)